### ISSUE TITLE
A few questions about long term operations

### ISSUE THREAD

**Issue by jinks** [dpc/rdedup#145]
I'm in the process of streamlining and consolidation my mess of different backup systems. (Currently I'm using rdedup, restic and borg to different degrees and on different machines, this is becoming unmaintainable.) While planning it all out a few questions have come up:

### Naming and rotation
Rdedup doesn't offer much assistance in backup naming and rotating out old backups. By the looks of it `list` gives an unsorted result and there is no way of determining the age of a backup besides encoding it in the name. What is a good way to name things so I can achieve a yearly-monthly-daily scheme and retiring and gc'ing old data? - Also: one repo per host or a shared repo with host based names? (This would be easier if rdedup had some kind of "bucket" or namespace system.)

### Security
Since I have some VPSs on shared systems I like the asymmetric encryption of rdedup, it at least prevents snooping. But, as far as I understand, attackers could still delete names and gc the repo. Is there a way to prevent this? My current plan is to contain those hosts to a separate repo, but that obviously reduces deduplication.

### Bandwidth use
One machine is connected to the backup server over a slow, metered link. What is the smarter play here in terms of bandwith use, deduplication and chunking. I can either mount the remote server over sshfs on the backup server and "pull" the data or run rdedup on the remote and push to the backup server via ssh. How much data does rdedup exchange with the repo? Less or more than a full read of the data to be backed up?

### "Backend" program
You suggest rdup for feeding in the data stream, but looking at [this comment](https://github.com/miekg/rdup/pull/32#issuecomment-372668201) as well as a few others, rdup's future is a bit up in the air. I've been evaluating libarchive's `bsdtar` as a replacement, using the pax format for tars so support xattrs, etc. Cpio would be another option, but I have little experience with it. Is there something that speaks for one or the other?

Phew, that's quite a novel. Feel free to redirect me somewhere else if this isn't the right place for these questions.

**Comment by dpc** [comment]
### Naming

`rdedup` core logic is in the library and `rdedup`-the-binary is just a small commandline api around it. I was focused on the core "hard" problems (library), and don't really have time and energy to try to turn the binary into more sophisticated backup solution. Everyone has their own particular ideas and needs.

You can quite easily write your own `rdedup` like binary on top of the `rdedup`-the-library, doing whatever schemes you wish (buckets, names, rotating) and not use the `rdedup`-the-binary. They would still remain compatible, because the underlying storage and accessing it, is the same and handled by the same library.



### Security

Yes. There's nothing that can be done about. I'm using `rclone` to sync my local copy to backblaze b2 that is cheap, and retains copy of the deleted files for a month or so, so I'm less afraid of loosing my backup.

### Bandwidth

I'm using `syncthing` to synchronize between hosts, and `rclone` to synchronize with backblaze b2. Rdedup is very careful about storage immutability and being sync-friendly, so it's actually very easy for software like `rsync`, Dropbox etc. to very cheaply synchronize data without any unnecessary data being sent.

#### "Backend" program

I would actually call it "frontend". :) . Rdedup comes with support for "backend" (storage: local, potentialy (not implemented yet) cloud services etc).

Anything that does not scramble the backed up fileset too much will work. There is some work being done by `rdedup` users to come up with a custom "file traverser" that would store data more optimally than `tar` and similiar.


Please join gitter channel and try asking around.

**Comment by jinks** [comment]
> You can quite easily write your own rdedup like binary on top of the rdedup-the-library

Looks like a good excuse to expand my rust knowledge beyond the "I can read it without getting an aneurysm" stage.

> I'm using syncthing to synchronize between hosts, and rclone to synchronize with backblaze b2. 

So you're keeping a local repo on each computer and then sync them to their "final destination" by external means? I'm not sure I have the storage space to do that in all cases. I'll have to do some experimenting.

**Comment by dpc** [comment]
https://docs.rs/rdedup-lib/3.0.0-0.2.0/rdedup_lib/

> So you're keeping a local repo on each computer and then sync them to their "final destination" by external means?

All computers are the final destination. They sync among each other. So on my 3 computer, I have 3 copies of my backup. On one of them, I additionally sync to backblaze.

Obviously, depending on needs and capacity, different setups might be created.

**Comment by jinks** [comment]
Alright, one final question:

If I understand that correctly, I can backup into a repo on hostA, then sync that repo to hostB, do a backup there and sync that back to hostA *without* fear of repo corruption despite the 2 rdedups not knowing about each other?

**Comment by dpc** [comment]
Correct. The sync and both operations can even be happening at the same time, and everything will be OK. Rdedup guarantees no two files will ever be created under the same name, with different content, and all operations are ordered/atomic so they don't confuse syncing tools.
